docs/data/data-grid/recipes-editing/BulkEditing.tsx

```diff
@@ -0,0 +1,206 @@
+        type: 'actions',
+        getActions: ({ id, row }) => {
+          return [
+            <GridActionsCellItem
+              icon={<RestoreIcon />}
+              label="Discard changes"
+              disabled={unsavedChangesRef.current.unsavedRows[id] === undefined}
+              onClick={() => {
+                apiRef.current.updateRows([
+                  unsavedChangesRef.current.rowsBeforeChange[id],
+                ]);
+                delete unsavedChangesRef.current.rowsBeforeChange[id];
+                delete unsavedChangesRef.current.unsavedRows[id];
+                setHasUnsavedRows(
+                  Object.keys(unsavedChangesRef.current.unsavedRows).length > 0,
+                );
+              }}
+            />,
+            <GridActionsCellItem
+              icon={<DeleteIcon />}
+              label="Delete"
+              onClick={() => {
+                unsavedChangesRef.current.unsavedRows[id] = {
+                  ...row,
+                  _action: 'delete',
+                };
+                if (!unsavedChangesRef.current.rowsBeforeChange[id]) {
+                  unsavedChangesRef.current.rowsBeforeChange[id] = row;
+                }
+                setHasUnsavedRows(true);
+                apiRef.current.updateRows([row]); // to trigger row render
+              }}
+            />,
+          ];
+        },
+      },
+      ...data.columns,
+    ];
+  }, [data.columns, unsavedChangesRef, apiRef]);
+
+  const processRowUpdate: NonNullable<DataGridProps['processRowUpdate']> = (
+    newRow,
+    oldRow,
+  ) => {
+    const rowId = newRow.id;
+
+    unsavedChangesRef.current.unsavedRows[rowId] = newRow;
+    if (!unsavedChangesRef.current.rowsBeforeChange[rowId]) {
+      unsavedChangesRef.current.rowsBeforeChange[rowId] = oldRow;
+    }
+    setHasUnsavedRows(true);
+    return newRow;
+  };
+
+  const discardChanges = () => {
+    setHasUnsavedRows(false);
+    Object.values(unsavedChangesRef.current.rowsBeforeChange).forEach((row) => {
+      apiRef.current.updateRows([row]);
+    });
+    unsavedChangesRef.current = {
+      unsavedRows: {},
+      rowsBeforeChange: {},
+    };
+  };
+
+  const saveChanges = async () => {
+    try {
+      // Persist updates in the database
+      setIsSaving(true);
+      await new Promise((resolve) => {
+        setTimeout(resolve, 1000);
+      });
+
+      setIsSaving(false);
+      const rowsToDelete = Object.values(
+        unsavedChangesRef.current.unsavedRows,
+      ).filter((row) => row._action === 'delete');
+      if (rowsToDelete.length > 0) {
+        rowsToDelete.forEach((row) => {
+          apiRef.current.updateRows([row]);
+        });
+      }
+
+      setHasUnsavedRows(false);
+      unsavedChangesRef.current = {
+        unsavedRows: {},
+        rowsBeforeChange: {},
+      };
+    } catch (error) {
+      setIsSaving(false);
+    }
+  };
+
+  return (
+    <div style={{ width: '100%' }}>
+      <div style={{ marginBottom: 8 }}>
```

Comment:
maybe use `Stack` and `Box` with the `sx` prop instead?

Reply:
I would rather keep regular DOM elements for styles as simple as above for the sake of simplicity 🙂